module/zfs/dsl_bookmark.c

```diff
@@ -116,84 +120,206 @@ dsl_bookmark_lookup(dsl_pool_t *dp, const char *fullname,
 	return (error);
 }
 
-typedef struct dsl_bookmark_create_redacted_arg {
-	const char	*dbcra_bmark;
-	const char	*dbcra_snap;
-	redaction_list_t **dbcra_rl;
-	uint64_t	dbcra_numsnaps;
-	uint64_t	*dbcra_snaps;
-	void		*dbcra_tag;
-} dsl_bookmark_create_redacted_arg_t;
+/*
+ * Validates that
+ * - bmark is a full dataset path of a bookmark (bookmark_namecheck)
+ * - source is a full path of a snaphot or bookmark
+ *   ({bookmark,snapshot}_namecheck)
+ *
+ * Returns 0 if valid, -1 otherwise.
+ */
+static int
+dsl_bookmark_create_nvl_validate_pair(const char *bmark, const char *source)
+{
+
+	if (bookmark_namecheck(bmark, NULL, NULL) != 0)
+		return (-1);
+
+	int is_bmark, is_snap;
+	is_bmark = bookmark_namecheck(source, NULL, NULL) == 0;
+	is_snap = snapshot_namecheck(source, NULL, NULL) == 0;
+	if (!is_bmark && !is_snap)
+		return (-1);
+
+	return (0);
+}
+
+/*
+ * Check that the given nvlist corresponds to the following schema:
+ *  { newbookmark -> source, ... }
+ * where
+ * - each pair passes dsl_bookmark_create_nvl_validate_pair
+ * - all newbookmarks are in the same pool
+ * - all newbookmarks have unique names
+ * - FIXME: need to check that newbookmark is never used as source in any other?
```

Comment:
I guess

```
fs#a -> fs@a
fs#b -> fs#a
```
would work because we run the list top-to-bottom.

```
fs#b -> fs#a
fs#a -> fs@a
```
would fail for the first pair with an error indicating `fs#a` does not exist, and have success with the second pair.

Not sure if we want to commit to the implementation detail that the list is executed top-to-bottom, and not in parallel.

What do you think?

Reply:
I agree it would be nice to disallow this.  We could check that all the sources exist, before creating any of them.

Reply:
`dsl_bookmark_create_nvl_validate`'s job is schema validation.

`dsl_bookmark_create_check` and `dsl_bookmark_create_redacted_check` already check existence for all sources.

I'll update the documentation of `dsl_bookmark_create_nvl_validate` and add the check to prevent the behavior described above.

Reply:
I think it would be fine for "schema validation" to not check that the sources exist (and not care about if they are trying to create the bookmark and use it in the same operation), just that their names are allowed strings, and they are all in the same pool.

Reply:
OK, so if I understand correctly, the current behavior is fine, i.e. it is ok that 
```
fs#a -> fs@a
fs#b -> fs#a
```
will create both bookmarks?

Reply:
- [x] extend pyzfs unittests for bookmark creation with missing sources
  to explicitly check that no bookmarks were created if a source is missing